src/Symfony/Component/Semaphore/Key.php

```diff
@@ -21,14 +21,16 @@
  */
 final class Key
 {
-    private string $resource;
     private int $limit;
     private int $weight;
     private ?float $expiringTime = null;
     private array $state = [];
 
-    public function __construct(string $resource, int $limit, int $weight = 1)
-    {
+    public function __construct(
+        private string $resource,
+        int $limit,
+        int $weight = 1,
```

Comment:
IIRC @OskarStark and me agreed on doing it this way for properties that are somehow processed in the constructor (here the error handling).

Reply:
But they're just validated, not changed. That should not prevent us from promoting them.

Reply:
My reasoning for not doing this is the following: When I scan the code for defined properties and we used CPP here, I would only recognise the `$weight` property, but I would probably not notice the validation happening here as I would stop reading the rest of the constructor then (if I am only interested in this particular property).

Reply:
I'd argue that if you move all boring assignments out of the constructor bodies, validation logic or consistency checks like this one stand out. It's a constructor with a non-empty body in a codebase with mostly empty constructor bodies.

Let's look at this from a different angle. Say, we have a class with five promoted properties and no additional logic in its constructor. In order to implement a change, I want to add a consistency check to that constructor. Following that logic, would I need to demote the properties involved in that check?

Here's one example of my HttpKernel PR. Would you have solved that one differently?

https://github.com/symfony/symfony/blob/54faed6c708ae95d8f829e0416e83fad41f657dc/src/Symfony/Component/HttpKernel/EventListener/RouterListener.php#L51-L64

* `$matcher` in that example is promoted, but still validated in the constructor.
* `$context` is not promoted because the property type is narrower than the corresponding constructor parameter type.

Reply:
I'm on the side of using CPP as much as possible, so here also

Reply:
The example from `RouterListener` convinced me. I have updated the code.